hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,567 @@
+        var feeList = new ArrayList<Fees>();
+        for (var entry : tokenAssociationsMap.entrySet()) {
+            final var tokenAssociateBody = TokenAssociateTransactionBody.newBuilder()
+                    .account(entry.getKey())
+                    .tokens(new ArrayList<>(entry.getValue()))
+                    .build();
+
+            final var syntheticTxn = TransactionBody.newBuilder()
+                    .tokenAssociate(tokenAssociateBody)
+                    .transactionID(feeContext.body().transactionID())
+                    .build();
+
+            feeList.add(feeContext.dispatchComputeFees(syntheticTxn, feeContext.payer()));
+        }
+
+        return combineFees(feeList);
+    }
+
+    private Fees combineFees(List<Fees> fees) {
+        long networkFee = 0, nodeFee = 0, serviceFee = 0;
+        for (var fee : fees) {
+            networkFee += fee.networkFee();
+            nodeFee += fee.nodeFee();
+            serviceFee += fee.serviceFee();
+        }
+        return new Fees(nodeFee, networkFee, serviceFee);
+    }
+
+    /**
+     * As part of pre-handle, token transfers in the transfer list are plausible.
+     *
+     * @param tokenID      The ID of the token we are transferring
+     * @param transfers    The transfers to check
+     * @param ctx          The context we gather signing keys into
+     * @param accountStore The account store to use to look up accounts
+     * @throws PreCheckException If the transaction is invalid
+     */
+    private void checkFungibleTokenTransfers(
+            @NonNull final TokenID tokenID,
+            @NonNull final List<AccountAmount> transfers,
+            @NonNull final PreHandleContext ctx,
+            @NonNull final ReadableAccountStore accountStore)
+            throws PreCheckException {
+        final var tokenStore = ctx.createStore(ReadableTokenStore.class);
+        final var tokenRelStore = ctx.createStore(ReadableTokenRelationStore.class);
+        // Fail if we have custom fees attached to the token
+        validateTruePreCheck(tokenHasNoCustomFeesPaidByReceiver(tokenID, tokenStore), INVALID_TRANSACTION);
+        // We're going to iterate over all the transfers in the transfer list. Each transfer is known as an
+        // "account amount". Each of these represents the transfer of fungible token INTO a single account or OUT of a
+        // single account.
+        for (final var accountAmount : transfers) {
+            // Given an accountId, we need to look up the associated account.
+            final var accountId = validateAccountID(accountAmount.accountIDOrElse(AccountID.DEFAULT), null);
+            final var account = accountStore.getAliasedAccountById(accountId);
+            final var isCredit = accountAmount.amount() > 0;
+            final var isDebit = accountAmount.amount() < 0;
+            if (account != null) {
+                if (isStakingAccount(ctx.configuration(), account.accountId()) && (isDebit || isCredit)) {
```

Comment:
We can remove `isDebit || isCredit` and just always throw here.

Reply:
Removed